kafka-client/src/main/java/dev/responsive/kafka/internal/stores/AbstractResponsiveStore.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2023 Responsive Computing, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dev.responsive.kafka.internal.stores;
+
+import static org.apache.kafka.streams.processor.internals.ProcessorContextUtils.asInternalProcessorContext;
+
+import dev.responsive.kafka.internal.db.CassandraClient;
+import dev.responsive.kafka.internal.utils.TableName;
+import java.util.Collection;
+import java.util.function.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.processor.StateStoreContext;
+import org.apache.kafka.streams.processor.internals.InternalProcessorContext;
+import org.apache.kafka.streams.query.Position;
+
+public abstract class AbstractResponsiveStore implements StateStore {
+
+  protected final TableName name;
+  protected final Position position; // TODO(IQ): update the position during restoration
+
+  // Initialized during #init
+  @SuppressWarnings("rawtypes")
+  protected InternalProcessorContext context;
+  protected TopicPartition changelog;
+  protected CassandraClient client;
+  protected ResponsiveStoreRegistry storeRegistry;
+  protected ResponsiveStoreRegistration registration;
+
+  private boolean open;
+
+  public AbstractResponsiveStore(final TableName name) {
+    this.name = name;
+    this.position = Position.emptyPosition();
+  }
+
+  @SuppressWarnings("rawtypes")
+  protected void init(
```

Comment:
another trick, by the way, to make sure that variables (especially non-final ones) remain under control of the super class is to pass them in to the methods you expect the subclass to override. 

Here's a quick example:

```java
class Base {
  private int foo;

  final void init() {
    foo = 3;
  }

  final int magic() {
    return doMagic(foo);
  }

  protected abstract int doMagic(final int foo);
}

class Sub {

  @Override
  protected int doMagic(final int foo) {
    return foo * 2;
  }

}
```

Generally what's nice here is that you can read each class totally independently of one another - here `doMagic` is a very clear formula, and in the super class `magic()` is clear that it delegates to some implementation, but they're both actually entirely independent.

Reply:
1. Good point about init/open, you're totally right I have the parent-child relationship completely backwards right now. Thanks for the feedback and suggestions, let me try again with a more principled approach
2. Regarding the `position` thing, it's certainly true I left the ownership/responsibility unclear, although that was somewhat intentional -- there's no contract that defines where/when it should be updated because I'm pretty sure we're not actually updating it everywhere it should be in the current implementation, and we'll need to do a thorough pass if/when we do want to enable IQ (see the TODO). But I'll try to get it into a better place for now